### ISSUE TITLE
Would it be possible to provide some ledger commands?

### ISSUE THREAD

**Issue by daraul** [bradyt/cone#65]
I've been thinking, it might be useful to include some ledger commands, and put their output on a page in cone. A balance command comes to mind immediately, but others may prove useful. I wouldn't expect one to recreate these functions in cone, but perhaps some version of the ledger binary could be included, and used instead?

**Comment by bradyt** [comment]
Sorry I haven't made time to write a response. I wanted to provide a lot of context for how we might consider your question. So instead, I'll just provide my opinion briefly.

I think the first question is, do we do this with some existing software out there, like ledger-cli, hledger, beancount, etc?

I really don't know much about porting these tools to Android. I think taskwarriorc2 somehow wraps a c++ tool into an Android app. And you can find on some wiki information about efforts to get Haskell programs on Android.

- https://wiki.haskell.org/Android
- https://bitbucket.org/kvorobyev/taskwarriorc2/src/master/

However, I'm hoping to port this app to iOS soon, and potentially other platforms. So I presume we might want a library that works across platforms, and it seems reasonable to attempt this in the current language, Dart. (I'm curious if we could get a prototype ledger library in F# started, and somehow get that working as a plugin for this Flutter front end.)

I guess if we implement in Dart, we could start with small steps, like just adding amounts, and the first feature to use it might just be showing the balancing amount in the hint text in the data entry form. Then we could start to think about some easy balance features, such as, what is the balance of all of today's transactions, and add more as we go.

**Comment by daraul** [comment]
It is actually because of taskwarriorc2's wiki mentioning including the taskwarrior binary somewhere that I got the idea that this might be possible at least in theory. 

On the one hand I am a fan of the idea of a purpose-built library that we use to keep things nice and clean, while supporting multiple platforms.

On the other, ledger is a dizzyingly powerful piece of software all on it's own, so trying to make a library mimics it is daunting to say the least, though I don't foresee cone ever having to support every single one of it's features. 

I suppose the best solution in the long run is to port the ledger-cli into Dart or F# as you mentioned. I'd be happy to help, though I have no experience with either language.

**Comment by bradyt** [comment]
On the granular level, to get a balance command, presumably we need a way to add two amounts. I'm not sure the code we have so far, is completely intentional about when to represent a quantity with a number, and when to represent it as a string.

I guess we'd simultaneously want to think about the big picture. We want a balance command. We'll probably want a structure that represents the tree of accounts, with amounts at each node. And then for each posting, we'll want to update that tree. We might start with an empty tree, and then create a path for each posting, so if the first posting we encounter is "A:B:C  3.00 XTS", then we would start some sort of tree for that, with maybe three nodes, and presumably put the 3.00 XTS at the leaf node.

If the next posting is "A:B:D  5.00 XTS", then we would put another child to B and put D there with 5.00 XTS.

As soon as we get to another posting about the A:B:C account, you see why we needed to implement a way to add two amounts. So if we encounter "A:B:C  -8.00 XTS", then we would go to the C node and combine the two amounts "3.00 XTS" and "-8.00 XTS". Depending on however we were representing the amount objects, this might be easy.

You'd also want to handle cases like "5.00 AAA" plus "7.00 BBB", which presumably should be stored as just an amount with two parts. This is usually when I start to go look at how hledger is doing things. I don't recall if I've asked or looked about this case.

**Comment by bradyt** [comment]
There's an ad-hoc cli tool in the cone_lib library, and sometimes I just add on a debug or print command to quickly iterate on changes. Or sometimes you can even uses the test files to quickly see how something is showing.

If you combine that with a tool like `entr`, you could get fast feedback on changes.

Sometimes developing on things in cone_lib, motivates having to shift things around in the cone_flutter directory, and that's fine.

**Comment by daraul** [comment]
Being able to report the balance of an account seems a sensible first step for a ledger library in cone, though we should keep in mind that when calculating the balance of a virtual account we should consider *all* postings to that account, while we should only consider *real* postings when calculating the balance of a real account.
Otherwise we'll end up having virtual account balance assertions working, while real account assertions reporting unbalanced. \

**Comment by daraul** [comment]
I just remembered I had done some research on ways to package a linux binary in an Android/Flutter application, I will update the issue description with some SOF answers I was able to find. They may prove useful if we decide to go that route.

**Comment by bradyt** [comment]
> Being able to report the balance of an account seems a sensible first step for a ledger library in cone, though we should keep in mind that when calculating the balance of a virtual account we should consider all postings to that account, while we should only consider real postings when calculating the balance of a real account.
>
> Otherwise we'll end up having virtual account balance assertions working, while real account assertions reporting unbalanced. \

Virtual accounts and balance assertions don't seem like things that need to be prioritized or blocking. I think trying first to just implement the full balance tree would immediately add a nice feature to an already working data entry app. We can work on additional features as we go.

Unless I misunderstood something. Do you have an example where a virtual account and balance assertion should stop cone in its tracks? Wouldn't it be fine initially to just ignore the balance assertion? For now, users can rely on their real computers for this. This app just brings a few features to mobile.

**Comment by daraul** [comment]
Of course, with a custom library to handle balance assertions there's no reason for any balance assertion to cause cone to misbehave. We could simply ignore balance assertions.

**Comment by daraul** [comment]
I also found this tutorial that may provide the best of both worlds, though this is from my very cursory skimming of the article, with no prior experience using flutter: https://flutter.dev/docs/development/platform-integration/c-interop

I'll include it in the issue description.

**Comment by bradyt** [comment]
Hmm, maybe I should add a section to the wiki. I'd been meaning to add a section on alternative workflows to dedicated installed apps (like ssh/mosh, termux, text editors, shorthand, web apps, etc.). Even using them in concert with the dedicated apps. I imagined adding items without need for stable connection, with cone, and then possibly syncing that to a VPS and serving via hledger-web, for access to calculations.

So maybe I could copy some of your research on what someone might consider in getting ledger-cli/hledger/beancount/etc engines working on Android or iOS. I forgot about that ffi, I don't remember why I didn't pursue it further. I think some of beancount's parser is implemented in C.

**Comment by daraul** [comment]
Yes, some specifics on getting ledger-cli setup within cone would be fantastic. I've tried reading the article, but I can't make heads or tails of it as yet.
